File: test/python/bench/sieve_primes.py
"""Mirror of test/jacl/bench/sieve_primes.jacl.

Count primes below N=2000 via trial division. Pure arithmetic, no
allocation. Expected output: 303.

NB: Python doesn't have a JACL-style early break here, so we use
`break` once we've found a divisor, matching the early-exit shape
JACL would have if it used break too. To keep the work apples-to-
apples we still scan i*i <= k either way.
"""


def run() -> int:
    n = 2000
    count = 0
    k = 2
    while k < n:
        is_p = True
        i = 2
        while i * i <= k:
            if k % i == 0:
                is_p = False
            i += 1
        if is_p:
            count += 1
        k += 1
    return count


if __name__ == "__main__":
    print(run())

File: test/clojure/bench/sieve_primes.clj
;; Mirror of test/jacl/bench/sieve_primes.jacl.
;;
;; Count primes below N=2000 via trial division. Pure arithmetic, no
;; allocation. Validates: 303.

(ns bench.sieve-primes)

(defn ^boolean prime? [^long n]
  (loop [i 2]
    (cond
      (> (* i i) n) true
      (zero? (rem n i)) false
      :else (recur (inc i)))))

(defn run []
  (loop [k 2 count 0]
    (if (< k 2000)
      (recur (inc k) (if (prime? k) (inc count) count))
      count)))
